Singleton widgets

ywwg opened the pull request:
Add support for "singleton" widgets in the skin system, and apply it to LateNight's library.

 WSingletonContainer defines widgets that should only be instantiated once
 but may appear in multiple places in a skin definition.  This is useful
 for complex widgets like the library, which are memory intensive. The
 container mostly looks like a special WidgetGroup which is defined in
 special ways.

 Usage:
 First, the Singleton container is defined, meaning it is described to the
 skin system by name, and what the singleton consists of.  This definition
 should be very early in the skin file.  Note that the singleton does not
 actually appear where it is defined.

 Example definition:
 `<DefineSingleton>
   <ObjectName>LibrarySingleton</ObjectName>
   <Layout>horizontal</Layout>
   <SizePolicy>me,me</SizePolicy>
   <Children>
     <Template src="skin:library.xml"/>
   </Children>
  </DefineSingleton>`

 The ObjectName is used to identify this singleton elsewhere in the skin
 files.

 Example usage:
 `<WidgetGroup>
    <ObjectName>SomeUiElement</ObjectName>
    <Layout>vertical</Layout>
    <SizePolicy>min,i</SizePolicy>
    <Children>
      <Singleton objectName="LibrarySingleton"/>
      ...
    </Children>
 </WidgetGroup>`

 The skin system sees the Singleton tag, and any time the containing
 group gets a show event, the Singleton widget is reparented to this location
 in the skin.  Note that if a Singleton is visible twice at the same time,
 behavior is undefined and could be crashy.

ywwg committed:
First attempt at creating only one Library widget and reparenting as the view changes.

ywwg committed:
more gross experimental work.  WLibrary is a QStackedWidget which has its own layout.

ywwg committed:
Figure out a layout remove/readd method that works

ywwg committed:
Merge branch 'master' into library-reparenting

ywwg committed:
just a note

ywwg committed:
Hook up to the correct signal.

Remove debugging and false starts

ywwg committed:
Finish Singleton implementation

ywwg committed:
remove some cruft, add a comment

ywwg committed:
fix comment

ywwg committed:
missing bracket

daschuer commented on src/skin/legacyskinparser.cpp:
```diff
@@ -1084,6 +1089,29 @@ QWidget* LegacySkinParser::parseCoverArt(QDomElement node) {
+void LegacySkinParser::parseSingletonDefinition(QDomElement node) {
+    // The actual singleton definition is identical to any other WWidgetGroup.
+    QWidget* child_group = parseWidgetGroup(node);
+    WSingletonContainer::defineSingleton(child_group->objectName(),
```
This should be the outer function to avoid instantiate (and leak) a duplicate.

daschuer replied:
Or just merge them, the function name imprecise anyway.

ywwg replied:
I don't think it's leaked, ownership of the pointer is passed to the SingletonContainer class.  But I do have to figure out destruction -- if a user loads another skin, and then reloads the first skin, bad things may happen.

rryan replied:
As written singletons are parented to the parent widget where they were defined in the skin hierarchy and when a singleton container claims it the widget is reparented -- so the widget can never leak since it always has a parent. It would be nice to save the work on creating the widget if a singleton is already defined but this is probably a minor case and an error on the part of the skin writer anyway.

daschuer commented on src/skin/legacyskinparser.cpp:
```diff
@@ -488,6 +489,10 @@ QList<QWidget*> LegacySkinParser::parseNode(QDomElement node) {
+    } else if (nodeName == "DefineSingleton") {
+        parseSingletonDefinition(node);
+    } else if (nodeName == "Singleton") {
+        result = wrapWidget(parseSingletonContainer(node));
```
The names are OK, I have just suspected it the other way round. We have no verbs in the other tag names. So you may think about this: 

<Singleton>  Where the singleton is instantiated. This would match with the other widgets like <Pushbutton> which is instantiated when named.  Alternative <SingletonInstance>  
And <SingletonContainer> or <SingeltonView> for the different views. 

Have you also thought about to get rid of the predefinition and just add a singleton property to the templates? 
This might be not that flexible, but much easier to use. 
<Template singleton= true>

ywwg replied:
your comment is hard to read because github is eating the brackets.

RJ didn't want to use the template solution because it is less flexible and doesn't allow for explicit widgetgroup sizing definitions.

I could do "SingletonDefinition" and "SingletonContainer" -- but I actually like that we don't use a verb anywhere else because the singleton is a very different type of skin item.

rryan replied:
```Singleton``` is particularly confusing because ```PushButton``` creates a push-button so you'd expect that ```Singleton``` creates a singleton. So I'd say put a View or Container on it or something to make it clear it's the thing that is showing the singleton. 

I agree with Daniel about the verbs #bikeshed.

daschuer commented on src/skin/legacyskinparser.cpp:
```diff
@@ -1084,6 +1089,29 @@ QWidget* LegacySkinParser::parseCoverArt(QDomElement node) {
+        qWarning() << "ERROR: Singleton" << objectName
+                   << "not found, probably about to crash.";
+    }
+    commonWidgetSetup(node, pContainer);
```
Uhhh... I have mixed feeling with this call here. The inner code assumes that this is a new widget. 
Does Qt allow two parents for a single widget?  I am afraid we need a proper passing from one parent to an other to be bullet proof.

ywwg replied:
In the error case above I should just return new QWidget()

A single widget does not have two parents.  There are two WidgetGroups created, however.  One is the container, which is often empty (this is the actual WSingletonContainer).  The other is the WidgetGroup that is the singleton, created in the defineSingleton call and owned by the singletoncontainer.  This other group is the one that is reparented around, inserted in containers as necessary.

ywwg replied:
note that there is in fact a new widget -- the SingletonContainer itself is a WidgetGroup that needs common setup.

rryan replied:
right, the setup is for hte container, not the singleton widget itself -- it's important to be able to style / configure the container and the widget itself

rryan commented on src/skin/legacyskinparser.cpp:
```diff
@@ -1084,6 +1089,29 @@ QWidget* LegacySkinParser::parseCoverArt(QDomElement node) {
 
+void LegacySkinParser::parseSingletonDefinition(QDomElement node) {
+    // The actual singleton definition is identical to any other WWidgetGroup.
+    QWidget* child_group = parseWidgetGroup(node);
```
probably should call ```child_group->hide()```. If no insertion point is visible to start with then the singleton will show up wherever it was defined in the skin tree.

rryan commented on src/skin/legacyskinparser.cpp:
```diff
@@ -1084,6 +1089,29 @@ QWidget* LegacySkinParser::parseCoverArt(QDomElement node) {
+QWidget* LegacySkinParser::parseSingletonContainer(QDomElement node) {
+    if (!node.hasAttribute("objectName")) {
+        qWarning() << "Need objectName attribute for Singleton tag.";
+        return new QWidget();
```
parseXXX methods can return NULL if they can't produce a widget

rryan commented on src/skin/legacyskinparser.cpp:
```diff
@@ -1084,6 +1089,29 @@ QWidget* LegacySkinParser::parseCoverArt(QDomElement node) {
+    QString objectName = node.attribute("objectName");
+    WSingletonContainer* pContainer =
+            WSingletonContainer::getSingleton(objectName, m_pParent);
+    if (pContainer == NULL) {
```
No need to crash :) just return NULL

rryan commented on src/widget/wsingletoncontainer.cpp:
```diff
@@ -0,0 +1,55 @@
+WSingletonContainer::WSingletonContainer(QString objectname, QWidget* pParent)
+        : WWidgetGroup(pParent), m_pWidget(NULL) {
+    WidgetMap::const_iterator widget_it = m_singletons.find(objectname);
+    DEBUG_ASSERT(widget_it != m_singletons.end());
```
Need to handle the case that the user chose an attribute name that doesn't exist. Not ok to just crash in release builds :P.

ywwg replied:
That case is handled in the factory function.  Since the constructor is private, this should never happen so this is a good place for a DEBUG_ASSERT (it would indicate some sort of crazy threading problem, so a crash is a good option)

rryan replied:
Hm, why not just pass the widget directly in the constructor? 

It's fine to keep the assert but future developers may not be aware of that invariant. It's best to design this class so that it doesn't segfault in unexpected conditions (i.e. check for null m_pWidget before dereferencing, don't dereference an end() iterator, etc.).

ywwg replied:
I just made that change

rryan commented on src/widget/wsingletoncontainer.cpp:
```diff
@@ -0,0 +1,55 @@
+    WidgetMap::const_iterator widget_it = m_singletons.find(objectname);
+    DEBUG_ASSERT(widget_it != m_singletons.end());
+
+    m_pWidget = *widget_it;
```
When the widget is deleted (once its parent is deleted) then other container's pointers to the widget will be junk. You can hook the QObject::destroyed signal to get a callback and void the pointer or you can use QPointer which does this for you.

rryan commented on src/widget/wsingletoncontainer.h:
```diff
@@ -0,0 +1,77 @@
+
+    virtual ~WSingletonContainer() { }
+
+    // We don't want to end up with badly-constructed containers, so only
```
Storing singletons in the static context isn't good: we can't create multiple skins at once (if we do in the future), they aren't automatically cleaned up when the skin is destroyed. Instead could you put this map in LegacySkinParser or SkinContext? That way the map only exists for as long as its needed (during the skin construction).

ywwg replied:
skincontext sounds like a good place to put it

rryan commented:
> Example definition:
> <DefineSingleton> <ObjectName>LibrarySingleton</ObjectName> <Layout>horizontal</Layout> <SizePolicy>me,me</SizePolicy> <Children> <Template src="skin:library.xml"/> </Children> </DefineSingleton>

Open question -- why make every singleton a widget group? 

Why not:

```xml
<DefineSingleton>
  <WidgetType>
    .. specific widget creation stuff ..
  </WidgetType>
</DefineSingleton>
```

ywwg commented:
a WidgetGroup is easy because I just pass the call down to parseWidgetGroup and I get all of that functionality for free.  If I were to pass down to parseNode(), that's tricky because that returns a list of widgets, not just one.  What if more than one element is returned (such as the case with a template)?  I definitely don't want to be reparenting a list of children.  I also don't want to have to throw some strange error message if there's a list, and I don't want to guess about what the sizepolicy should be if I pack the items into a group.

rryan commented:
Yea, that's true. WidgetStack just takes the first widget and prints an error:

https://github.com/mixxxdj/mixxx/blob/master/src/skin/legacyskinparser.cpp#L622

ywwg committed:
Move SingletonWidget map into SkinContext for easier lifetime management.

Also address other notes.

ywwg commented:
Addressed notes.  I feel like it's weird how the map is mutated, so I'm going to move it into its own class.

ywwg committed:
Make the SingletonMap its own class

ywwg commented:
phew ok now it's not quite so weird and functional :)

ywwg committed:
style nit

ywwg committed:
fix comments

rryan commented on src/widget/wsingletoncontainer.cpp:
```diff
@@ -0,0 +1,48 @@
+
+void WSingletonContainer::showEvent(QShowEvent* event) {
+    Q_UNUSED(event);
+    QWidget* parent = m_pWidget->parentWidget();
```
could be null

rryan commented on src/widget/wsingletoncontainer.cpp:
```diff
@@ -0,0 +1,48 @@
+}
+
+void SingletonMap::defineSingleton(QString objectName, QWidget* widget) {
+    WidgetMap::const_iterator widget_it = m_singletons.find(objectName);
```
might be more straightforward/readable to call contains?

rryan commented on src/skin/legacyskinparser.cpp:
```diff
@@ -1084,6 +1089,30 @@ QWidget* LegacySkinParser::parseCoverArt(QDomElement node) {
+    child_group->hide();
+}
+
+QWidget* LegacySkinParser::parseSingletonContainer(QDomElement node) {
```
Why does SingletonMap instantiate SingletonContainer? If it didn't then you could get rid of this method entirely and use parseStandardWidget. (also the widget's setup() method is never called which is conventional -- I could see why that would be intentional because you don't want the skinner to be able to specify a layout but OTOH you might want to specify a size constraint or BackPath for a SingletonContainer).

ywwg replied:
The idea is that the Map is in charge of managing the Containers.  The constructor of the container takes the widget being pointed at, and I really don't want callers to have to think about getting the pointer out of the map and then passing it to the container.  If that was the case, then callers are suddenly responsible for managing their own maps and could get into trouble.  I was even thinking of making the Container constructor still private, and making the Map a friend class so that people couldn't subvert the design and try to construct their own containers.

The idea is to keep the interface too simple to mess up -- tell the map "here is a new singleton," and later "give me the instance of the singleton with this name.

rryan replied:
> The idea is to keep the interface too simple to mess up -- tell the map "here is a new singleton," and later "give me the instance of the singleton with this name.

(sorry for continuing to hammer on such a minor thing)

It's "here is a new singleton," and later "give me the instance of the singleton with this name and also create a SingletonContainer without setting it up in the way widgets are normally set up and create it outside of LegacySkinParser which was previously the one-stop-shop to see where widgets are created." 

I understand wanting it to be simple but I'd say having more widgets use parseStandardWidget is simpler for the skin system overall because you don't have to read the code to know which widget types get event filters bound, which widget types get their Init method called (WSingletonContainer doesn't currently), which widget types get setup called, and in what order all of those methods gets called. This has caused headaches in the past (every widget parse method did things in its own order or in its own non-standard way) and I combined as many as possible into parseStandardWidget when I was pretty sure the non-standardness was accidental and not important. 

Ok I'm done -- feel free to keep it or change it. If you keep it the way it is please copy parseStandardWidget and add a note to keep it in sync with parseStandardWidget. :)

ywwg replied:
So I made this change in a separate branch as an experiment:  https://github.com/ywwg/mixxx/pull/4/files

I see what you mean about making the parsing standard, but there are a few things I don't like:

* SingletonContainer is now tightly coupled with SkinContext, with a weird cyclic dependency that means they can't be separated
* It's possible to have a badly-constructed container with null layout and widget, since that information can't be constructed until setup() time.

But I don't know how much of that is this specific widget, and how much is the skin design in general.  I guess my last-ditch argument is "well this isn't a standard widget ¯\_(ツ)_/¯" but that's pretty lame.  What do you think?

ywwg replied:
ah, so I may have found the compelling reason to do it this way.  I realized that I also have the spinnies in a widgetstack, so there are 2 spinnies for each deck.  In trying to make them singletons, I needed to be able to specify a different size for each spinny inside the SingletonContainer tag.  If we do standard widget parsing, we get the Size tag for free.

I also moved ObjectName into a tag so that variables can be used (variable parsing doesn't work in an attribute)

ywwg committed:
small notes

ywwg committed:
Convert WSingletonContainer to something that can be parsed with parseStandardWidget

ywwg committed:
Try to redo spinny with singleton

ywwg committed:
Merge branch 'singleton-widgets-spinny' into singleton-widgets-standard-spinny

Conflicts:
	src/skin/legacyskinparser.cpp

ywwg committed:
clean up now that it works

ywwg committed:
whitespace

ywwg commented:
Hm I think I have also convinced myself not to make the definition automatically a widgetgroup -- that just got in the way with the spinny work.

ywwg committed:
Make SingletonDefinition not a container itself.
Fix scope-dangerous variable name in WidgetStack code.

ywwg committed:
change comments

ywwg commented:
PTAL -- I basically ended up adopting all of the suggestions you made, even though I still have Opinions about the coupling between skincontext and widget setup.

ywwg committed:
Merge branch 'master' into singleton-widgets

ywwg committed:
oops

ywwg committed:
Fix library not appearing at smallest size if disabled at larger sizes (lp:#1413419)
move Vinyl button

ywwg commented:
friendly ping

daschuer commented on src/widget/wsingletoncontainer.cpp:
```diff
@@ -0,0 +1,61 @@
+}
+
+QWidget* SingletonMap::getSingletonWidget(QString objectName) const {
+    WidgetMap::const_iterator widget_it = m_singletons.find(objectName);
```
You can use her just:
m_singletons.value(objectName) 
or more explicit  
m_singletons.value(objectName, NULL)

daschuer commented on src/widget/wsingletoncontainer.h:
```diff
@@ -0,0 +1,90 @@
+
+class SingletonMap {
+  public:
+    typedef QMap<QString, QWidget*> WidgetMap;
```
This is only used Rarely, but hides implementation details where it is used.
(no meaningful pop-ups in Eclipse)  So I would prefer to not us a typedef here.

daschuer commented on src/skin/legacyskinparser.cpp:
```diff
@@ -1084,6 +1089,59 @@ QWidget* LegacySkinParser::parseCoverArt(QDomElement node) {
+    QDomNode childrenNode = m_pContext->selectNode(node, "Children");
+    if (childrenNode.isNull()) {
+        SKIN_WARNING(node, *m_pContext)
+                << "SingletonDefinition requires a Children tag with one child";
```
at least on child .. ?

daschuer replied:
Ah no. We support only one child. What is the reason to call the node "Children"?

ywwg replied:
For consistency.  See WidgetStack, which does the same thing.

daschuer commented on src/skin/legacyskinparser.cpp:
```diff
@@ -1084,6 +1089,59 @@ QWidget* LegacySkinParser::parseCoverArt(QDomElement node) {
+    // Descend chilren, taking the first valid element.
+    QDomNode child_node;
+    QDomNodeList children = childrenNode.childNodes();
+    for (int i = 0; i < children.count(); ++i) {
```
this can be a foreach loop

daschuer replied:
No. check here for count() == 1 and warn otherwise.

ywwg replied:
no, because the first node might not be an element. this loops through until it finds an element.

ywwg replied:
foreach doesn't work, there's no const_iterator for childNodes()

daschuer commented on src/skin/legacyskinparser.cpp:
```diff
@@ -1084,6 +1089,59 @@ QWidget* LegacySkinParser::parseCoverArt(QDomElement node) {
+
+    if (child_node.isNull()) {
+        SKIN_WARNING(node, *m_pContext)
+                << "SingletonDefinition child node is NULL";
```
"SingletonDefinition \"Children\"  node is empty"

daschuer commented on src/widget/wsingletoncontainer.cpp:
```diff
@@ -0,0 +1,61 @@
+    }
+}
+
+void SingletonMap::defineSingleton(QString objectName, QWidget* widget) {
```
insertSingleton()

daschuer commented on src/skin/skincontext.h:
```diff
@@ -79,6 +80,14 @@ class SkinContext {
     QDebug logWarning(const char* file, const int line, const QDomNode& node) const;
 
+    void defineSingleton(QString objectName, QWidget* widget) {
+        return m_pSingletons->defineSingleton(objectName, widget);
```
insertSingleton()

daschuer commented on src/widget/wsingletoncontainer.cpp:
```diff
@@ -0,0 +1,61 @@
+    if (!context.hasNode(node, "ObjectName")) {
+        SKIN_WARNING(node, context)
+                << "Need objectName attribute for Singleton tag";
+        return;
```
You allow to define a singelton with an empty name, but you return here. One path should be changed.

daschuer commented on src/widget/wsingletoncontainer.cpp:
```diff
@@ -0,0 +1,61 @@
+                << "Need objectName attribute for Singleton tag";
+        return;
+    }
+    QString objectName = context.selectString(node, "ObjectName");
```
Not related to this PR, but it looks like (not tested) that selectString() returns any other sting if "ObjectName" is not found. 
If we change it to returning an empty string in that case we can combine this selectString and hasNode to save on XML loop.

ywwg replied:
yeah definitely seems like a bug.

daschuer commented on src/widget/wsingletoncontainer.cpp:
```diff
@@ -0,0 +1,61 @@
+void WSingletonContainer::showEvent(QShowEvent* event) {
+    Q_UNUSED(event);
+    if (m_pWidget) {
+        QWidget* parent = m_pWidget->parentWidget();
```
Some comments would be helpfull here.
Does it steel the widget from Its original parent ... ?

ywwg committed:
Merge branch 'master' into singleton-widgets

ywwg committed:
Singleton Widgets: code review notes.

Also avoid an unnecessary reparent if the widget
already has the correct parent.

ywwg commented:
notes addressed

daschuer commented:
Thank you. LGTM
